--- ../src-base/minecraft/net/minecraft/client/renderer/chunk/RenderChunk.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/chunk/RenderChunk.java
@@ -143,13 +143,15 @@
             p_178581_4_.func_178540_f().unlock();
         }
 
+        if (cadiboo.renderchunkrebuildchunkhooks.hooks.RenderChunkRebuildChunkHooksHooks.onRebuildChunkEvent(field_178589_e, field_189564_r, p_178581_4_, compiledchunk, this.field_178586_f, p_178581_1_, p_178581_2_, p_178581_3_)) return;
         VisGraph lvt_9_1_ = new VisGraph();
         HashSet lvt_10_1_ = Sets.newHashSet();
 
         if (!this.field_189564_r.func_72806_N())
         {
             ++field_178592_a;
-            boolean[] aboolean = new boolean[BlockRenderLayer.values().length];
+            final cadiboo.renderchunkrebuildchunkhooks.event.RebuildChunkEvent.RebuildChunkBlocksEvent rebuildBlocksEvent = cadiboo.renderchunkrebuildchunkhooks.hooks.RenderChunkRebuildChunkHooksHooks.onRebuildChunkBlocksEvent(this.field_178589_e, this.field_189564_r, p_178581_4_, compiledchunk, BlockPos.func_177975_b(blockpos, blockpos1), Minecraft.func_71410_x().func_175602_ab(), this.field_178586_f, p_178581_1_, p_178581_2_, p_178581_3_, lvt_10_1_, lvt_9_1_);
+            final boolean[] aboolean = rebuildBlocksEvent.getUsedBlockRenderLayers();
             BlockRendererDispatcher blockrendererdispatcher = Minecraft.func_71410_x().func_175602_ab();
 
             for (BlockPos.MutableBlockPos blockpos$mutableblockpos : BlockPos.func_177975_b(blockpos, blockpos1))
@@ -195,8 +199,22 @@
                         this.func_178573_a(bufferbuilder, blockpos);
                     }
 
-                    aboolean[j] |= blockrendererdispatcher.func_175018_a(iblockstate, blockpos$mutableblockpos, this.field_189564_r, bufferbuilder);
+                    if (!rebuildBlocksEvent.isCanceled()) {
+                        final cadiboo.renderchunkrebuildchunkhooks.event.RebuildChunkEvent.RebuildChunkBlockEvent rebuildBlockEvent = cadiboo.renderchunkrebuildchunkhooks.hooks.RenderChunkRebuildChunkHooksHooks.onRebuildChunkBlockEvent(this.field_178589_e, this.field_189564_r, p_178581_4_, compiledchunk, blockrendererdispatcher, iblockstate, blockpos$mutableblockpos, bufferbuilder, this.field_178586_f, p_178581_1_, p_178581_2_, p_178581_3_, lvt_10_1_, lvt_9_1_);
+                        if (rebuildBlockEvent.isCanceled()) {
+                            for (final BlockRenderLayer blockrenderlayer : BlockRenderLayer.values()) {
+                                if (rebuildBlockEvent.getUsedBlockRenderLayers()[blockrenderlayer.ordinal()]) {
+                                    compiledchunk.func_178486_a(blockrenderlayer);
+                                }
+                            }
+                        } else {
+                            aboolean[j] |= blockrendererdispatcher.func_175018_a(iblockstate, blockpos$mutableblockpos, this.field_189564_r, bufferbuilder);
+                        }
+                    }
                 }
                 }
                 net.minecraftforge.client.ForgeHooksClient.setRenderLayer(null);
             }
 
             for (BlockRenderLayer blockrenderlayer : BlockRenderLayer.values())